src/components/projectconfig/configmgr.cpp

```diff
@@ -0,0 +1,399 @@
+#include "configmgr.h"
+#include <QtCore/QDir>
+#include <QtCore/QFile>
+#include <QtCore/QJsonArray>
+#include <QtCore/QJsonDocument>
+#include <QtCore/QJsonObject>
+
+#define MODULID_MASK 0xFF
+
+ConfigMgr::ConfigMgr()
+{
+}
+
+bool ConfigMgr::getAndCheckId(const QJsonObject& json, unsigned int& id)
+{
+    if (checkData(json, "Id", QJsonValue::Double) == false) {
+        return false;
+    }
+
+    id = json["Id"].toInt();
+    // Check range
+    if ((id <= Config::ModulType::MT_NULL) || (id >= Config::ModulType::MT_MAX)) {
+        cds_error("ConfigMgr::getAndCheckId - Id is out of range");
+        return false;
+    }
+    return true;
+}
+
+bool ConfigMgr::getAndCheckProperties(const QJsonObject& json, QJsonArray& arr)
+{
+    if (json.contains("properties") == false) {
+        cds_error("ConfigMgr::getAndCheckProperties - properties missing");
+        return false;
+    }
+    if (json["properties"].isArray() == false) {
+        cds_error("ConfigMgr::getAndCheckProperties - properties is not array");
+        return false;
+    }
+    arr = json["properties"].toArray();
+    return true;
+}
+
+bool ConfigMgr::checkData(const QJsonObject& json, const char* name, QJsonValue::Type type)
+{
+    if (json.contains(name) == false) {
+        cds_debug("ConfigMgr::checkValue - {} does not contain in object", name);
+        return false;
+    }
+    switch (type) {
+    case QJsonValue::Bool:
+        if (json[name].isBool() == false) {
+            cds_error("ConfigMgr::checkValue - {} format problem - bool is required", name);
+            return false;
+        }
+        break;
+
+    case QJsonValue::Double:
+        if (json[name].isDouble() == false) {
+            cds_error("ConfigMgr::checkValue - {} format problem - double is required", name);
+            return false;
+        }
+        break;
+
+    case QJsonValue::String:
+        if (json[name].isString() == false) {
+            cds_error("ConfigMgr::checkValue - {} format problem - string is required", name);
+            return false;
+        }
+        break;
+
+    case QJsonValue::Array:
+        if (json[name].isArray() == false) {
+            cds_error("ConfigMgr::checkValue - {} format problem - array is required", name);
+            return false;
+        }
+        break;
+
+    case QJsonValue::Object:
+        if (json[name].isObject() == false) {
+            cds_error("ConfigMgr::checkValue - {} format problem - object is required", name);
+            return false;
+        }
+        break;
+    default:
+        cds_error("ConfigMgr::checkValue - {} format problem - object is required", name);
```

Comment:
...what?   
Doesn't default mean someone called the function with a completely unexpected value for "type"?  It sounds to me more like a "this should not happen" situation and the action you take should indicate that.